### PR TITLE
SUBMARINE-597. Support for SSH based git sync mode

### PR DESCRIPTION
**Pull request by manirajv06**
### What is this PR for?
Clone the source code from git repository using SSH based access. /code dir is the location to store the source code. SSH based access requires SSH host private key and known hosts to be passed through /etc/secret to K8 git sync init container process.

### What type of PR is it?
Improvement 

### Todos
* [ ] - Task

### How should this be tested?
* First time? Setup Travis CI as described on https://submarine.apache.org/contribution/contributions.html#continuous-integration
* Strongly recommended: add automated unit tests for any new or changed behavior
* Outline any manual steps to test the PR here.

### Questions:
* Does the licenses files need update? Yes/No
* Is there breaking changes for older versions? Yes/No
* Does this needs documentation? Yes/No

### CODE REVIEW

**Review by liuxunorg — COMMENTED** [review]
hi, @manirajv06  There are 2 problems with this PR：
1. We need to explain what this PR contains in the description of [SUBMARINE-597](https://issues.apache.org/jira/browse/SUBMARINE-597) and PR.
2. In the paragraph `What is this PR for?` is not put a link to [SUBMARINE-597](https://issues.apache.org/jira/browse/SUBMARINE-597), it should be the content of [SUBMARINE-597](https://issues.apache.org/jira/browse/SUBMARINE-597).

**Comment by manirajv06** [comment]
@wangdatan @tangzhankun Can you please take a look?

We will need to add the documentation to describe the steps to create secrets as prerequisite step. Please refer https://github.com/kubernetes/git-sync/blob/master/docs/ssh.md for more details.

For example,

1. To capture ssh host key of repository server, use the following command:

ssh-keyscan github.com > /tmp/known_hosts

2. To create the secret for corresponding host server in K8

kubectl --kubeconfig /Users/mani/.kube/kind-config-kind create secret generic git-creds --from-file=ssh=/Users/mani/.ssh/id_rsa --from-file=known_hosts=/tmp/known_hosts

**Comment by manirajv06** [comment]
@liuxunorg Taken care

**Comment by liuxunorg** [comment]
> @liuxunorg Taken care

I helped you update the content of the PR. Please pay attention to my changes.

**Comment by xunliu** [comment]
@manirajv06 I have two question:
1. This PR allows users to use ssh to operate the git repository in k8s?
2. Can you attach a screenshot of the running effect to PR `Screenshots (if appropriate)`?

File: submarine-server/server-submitter/submitter-k8s/src/main/java/org/apache/submarine/server/submitter/k8s/parser/ExperimentSpecParser.java [hunk 1]
```diff
@@ -182,18 +186,41 @@ private static V1PodTemplateSpec parseTemplateSpec(
...
-            .get(0).getName();
-        String path = podSpec.getInitContainers().get(0).getVolumeMounts()
-            .get(0).getMountPath();
-
-        V1VolumeMount mount = new V1VolumeMount();
-        mount.setName(volumeName);
-        mount.setMountPath(path);
 
+        List<V1VolumeMount> initContainerVolumeMounts =
+            podSpec.getInitContainers().get(0).getVolumeMounts();
```
Comment (jiwq) [inline submarine-server/server-submitter/submitter-k8s/src/main/java/org/apache/submarine/server/submitter/k8s/parser/ExperimentSpecParser.java:191]: Why get the first item?

**Review by jiwq — COMMENTED** [review]
Thanks @manirajv06 for the work. I left some questions inline.

**Comment by manirajv06** [comment]
> 1. This PR allows users to use ssh to operate the git repository in k8s?

Source code available at git hub repos can be pulled into /code destination directory in k8 container using ssh based access. Earlier, we had support for "http" based access and now for "ssh" based access.

> 2. Can you attach a screenshot of the running effect to PR `Screenshots (if appropriate)`?

I don't have screenshots. It pulls the code from git hub repos and place it in /code destination directory.

**Comment by manirajv06** [comment]
> I helped you update the content of the PR. Please pay attention to my changes.

Thanks.

File: submarine-server/server-submitter/submitter-k8s/src/main/java/org/apache/submarine/server/submitter/k8s/parser/ExperimentSpecParser.java (hunk 1 above)
Comment (manirajv06) [reply submarine-server/server-submitter/submitter-k8s/src/main/java/org/apache/submarine/server/submitter/k8s/parser/ExperimentSpecParser.java:191]: There are 2 init containers. 1. for git sync process and 2. for environment setup. 0th init container has been created for git code sync process. As part of git sync, it is required to get some properties of corresponding init container and set the same in pod spec level as well. Hence, this step.

**Comment by xunliu** [comment]
@manirajv06 I have a question:
1. In the future, in the k8s container, can I also use `ssh` to operate github through `notebook` or `terminal`? This requires sharing some private key information.

Maybe you need to create some jira to achieve the above functions. Can you raise this question at the next community meeting, can we discuss it together?

**Comment by manirajv06** [comment]
> @manirajv06 I have a question:
> 
> 1. In the future, in the k8s container, can I also use `ssh` to operate github through `notebook` or `terminal`? This requires sharing some private key information.
> 
> Maybe you need to create a new jira to support this new feature. Can you raise this question at the next community meeting, can we discuss it together?

Yes, this requires sharing private key info. Ideally, we will need to get this info through Web UI to create the secrets in K8s and use it while creating init containers. May be I can write a design doc and run it through with others. This requires co-ordination with folks working on UI. cc @tangzhankun  @wangdatan 

For now, secrets can be created through k8 commands.

**Review by xunliu — APPROVED** [review]
@manirajv06 Thank you for contribution this featue.
LGTM

**Pull request closed without merging**